File: ruby/test_simple.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative 'lib/sensitive_string'

puts "=== Ruby SensitiveString Tests ===\n\n"

# Test 1: to_s shows hash
password = SensitiveString.new('my-secret-password')
result = password.to_s
raise "Expected hash format" unless result.start_with?('sha256:')
raise "Leaked plaintext!" if result.include?('my-secret-password')
puts "✅ Test 1: to_s shows hash"

# Test 2: value returns plaintext
raise "value() failed" unless password.value == 'my-secret-password'
puts "✅ Test 2: value returns plaintext"

# Test 3: String interpolation
message = "Password: #{password}"
raise "Leaked in interpolation!" if message.include?('my-secret-password')
puts "✅ Test 3: String interpolation shows hash"

# Test 4: inspect
inspect_str = password.inspect
raise "inspect leaked!" if inspect_str.include?('my-secret-password')
puts "✅ Test 4: inspect shows hash"

# Test 5: JSON
require 'json'
data = { password: password }
json = JSON.generate(data)
raise "JSON leaked!" if json.include?('my-secret-password')
puts "✅ Test 5: JSON serialization shows hash"

# Test 6: Equality
password2 = SensitiveString.new('my-secret-password')
raise "Equality failed" unless password == password2
puts "✅ Test 6: Equality works"

# Test 7: Helper methods
raise "extract_value failed" unless SensitiveString.extract_value(password) == 'my-secret-password'
raise "extract_value failed" unless SensitiveString.extract_value('plain') == 'plain'
puts "✅ Test 7: Helper methods work"

# Test 8: Consistent hash
raise "Hash inconsistent" unless password.to_s == password2.to_s
puts "✅ Test 8: Consistent hash for same value"

puts "\n✨ All tests passed!"

File: php/test_simple.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/src/SensitiveString.php';

use SensitiveString\SensitiveString;

echo "=== PHP SensitiveString Tests ===\n\n";

// Test 1: __toString shows hash
$password = new SensitiveString('my-secret-password');
$result = (string)$password;
assert(str_starts_with($result, 'sha256:'), "Expected hash format");
assert(!str_contains($result, 'my-secret-password'), "Leaked plaintext!");
echo "✅ Test 1: __toString shows hash\n";

// Test 2: getValue returns plaintext
assert($password->getValue() === 'my-secret-password', "getValue() failed");
echo "✅ Test 2: getValue returns plaintext\n";

// Test 3: value property returns plaintext
assert($password->value === 'my-secret-password', "value property failed");
echo "✅ Test 3: value property returns plaintext\n";

// Test 4: String interpolation
$message = "Password: $password";
assert(!str_contains($message, 'my-secret-password'), "Leaked in interpolation!");
echo "✅ Test 4: String interpolation shows hash\n";

// Test 5: JSON
$data = ['password' => $password];
$json = json_encode($data);
assert(!str_contains($json, 'my-secret-password'), "JSON leaked!");
assert(str_contains($json, 'sha256:'), "JSON doesn't show hash!");
echo "✅ Test 5: JSON serialization shows hash\n";

// Test 6: Helper methods
assert(SensitiveString::extractValue($password) === 'my-secret-password', "extract_value failed");
assert(SensitiveString::extractValue('plain') === 'plain', "extract_value failed");
echo "✅ Test 6: Helper methods work\n";

// Test 7: Consistent hash
$password2 = new SensitiveString('my-secret-password');
assert((string)$password === (string)$password2, "Hash inconsistent");
echo "✅ Test 7: Consistent hash for same value\n";

// Test 8: echo
ob_start();
echo $password;
$output = ob_get_clean();
assert(!str_contains($output, 'my-secret-password'), "echo leaked!");
echo "✅ Test 8: echo shows hash\n";

echo "\n✨ All tests passed!\n";

